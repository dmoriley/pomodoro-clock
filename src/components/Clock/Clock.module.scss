.root {
  padding: var(--s-size) var(--m-size);
  display: inline-block;
  border: 0.6rem solid lighten(#192944, 10%);
  border-radius: 20%;

  * {
    margin: 0;
  }
}

.time {
  font-family: Digital;
  font-size: 8rem;
  line-height: 1;
}