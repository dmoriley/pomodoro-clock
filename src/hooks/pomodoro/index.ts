export { default as usePomodoro } from './pomodoro.hook';
