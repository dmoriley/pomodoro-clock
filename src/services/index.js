/* PLOP_INJECT_IMPORT */
import time from './time';

export {
  /* PLOP_INJECT_EXPORT */
	time,
}