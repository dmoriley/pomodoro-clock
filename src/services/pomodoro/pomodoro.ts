import { BehaviorSubject } from 'rxjs';

export const SessionType = {
  session: 1,
  break: 2,
};

export const defaultBreakLength = 5; // minutes
export const defaultSessionLength = 20; //minutes
export const defaultTime = defaultSessionLength * 60;
export const defaultType = SessionType.session;

const createPomodoro = () => {
  let interval;

  const isOn = new BehaviorSubject(false);
  const type = new BehaviorSubject(defaultType);
  const sessionLength = new BehaviorSubject(defaultSessionLength);
  const breakLength = new BehaviorSubject(defaultBreakLength);
  const time = new BehaviorSubject(defaultTime); // time displayed on the clock

  const start = (callback, delay = 1000) => {
    isOn.next(true);
    interval = setInterval(callback, delay);
  };

  const pause = () => {
    isOn.next(false);
    clearInterval(interval);
  };

  const reset = () => {
    pause();
    type.next(defaultType);
    sessionLength.next(defaultSessionLength);
    breakLength.next(defaultBreakLength);
    time.next(defaultTime);
  };

  const playPause = () => {
    if (!isOn.value) {
      isOn.next(true);
      start(() => {
        if (time.value !== 0) {
          time.next(time.value - 1);
        } else if (type.value === SessionType.session) {
          new Audio('./sounds/success.mp3').play();
          type.next(SessionType.break);
          time.next(breakLength.value * 60);
        } else {
          new Audio('./sounds/doubleBeep.mp3').play();
          type.next(SessionType.session);
          time.next(sessionLength.value * 60);
        }
      });
    } else {
      // pause and turn off timer
      pause();
    }
  };

  const breakIncrement = () => {
    if (!isOn.value && breakLength.value < 60) {
      breakLength.next(breakLength.value + 1);
    }
  };

  const breakDecrement = () => {
    if (!isOn.value && breakLength.value > 1) {
      breakLength.next(breakLength.value - 1);
    }
  };

  const sessionIncrement = () => {
    if (!isOn.value && sessionLength.value < 60) {
      sessionLength.next(sessionLength.value + 1);
      time.next(sessionLength.value * 60);
    }
  };

  const sessionDecrement = () => {
    if (!isOn.value && sessionLength.value > 1) {
      sessionLength.next(sessionLength.value - 1);
      time.next(sessionLength.value * 60);
    }
  };

  return {
    isOn$: isOn.asObservable(),
    getIsOn() {
      return isOn.value;
    },
    type$: type.asObservable(),
    getType() {
      return type.value;
    },
    sessionLength$: sessionLength.asObservable(),
    getSessionLength() {
      return sessionLength.value;
    },
    breakLength$: breakLength.asObservable(),
    getBreakLength() {
      return breakLength.value;
    },
    time$: time.asObservable(),
    getTime() {
      return time.value;
    },
    reset,
    playPause,
    breakIncrement,
    breakDecrement,
    sessionIncrement,
    sessionDecrement,
  };
};

const singleton = createPomodoro();
Object.freeze(singleton);

export default singleton;
