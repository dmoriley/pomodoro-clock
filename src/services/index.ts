/* PLOP_INJECT_IMPORT */
import pomodoro from './pomodoro';
import time from './time';

export {
  /* PLOP_INJECT_EXPORT */
  pomodoro,
  time,
};
