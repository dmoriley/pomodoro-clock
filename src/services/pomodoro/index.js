import pomodoro, { SessionType } from './pomodoro';

export default pomodoro;
export { SessionType };
