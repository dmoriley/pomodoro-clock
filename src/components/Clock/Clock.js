import React from 'react';
import PropTypes from 'prop-types';
import styles from './Clock.module.scss';
import {time as timeFormat} from '../../services';
import { ClockState } from '../ClockContainer';

const Clock = ({time, status, power}) => {
  
  return (
    <div className={`${styles.root} ${!power ? styles.off : status === ClockState.session ? styles.session : styles.break }`}>
      {
        status === ClockState.session 
        ? <h2>Session</h2>
        : <h2>Break</h2>
      }
      
      <p className={styles.time}>{timeFormat.formatSeconds(time)}</p>
    </div>
  );
};

Clock.propTypes = {
  time: PropTypes.number.isRequired,
  status: PropTypes.number.isRequired,
  power: PropTypes.bool.isRequired
};

export default Clock;