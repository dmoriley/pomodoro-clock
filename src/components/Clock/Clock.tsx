import React from 'react';
import styles from './Clock.module.scss';
import { time as timeFormat } from '../../services';
import { SessionType } from '../../services/pomodoro';

type IProps = {
  time: number;
  status: number;
  power: boolean;
};

const Clock: React.FC<IProps> = ({ time, status, power }) => {
  return (
    <div
      className={`${styles.root} ${
        !power
          ? styles.off
          : status === SessionType.session
          ? styles.session
          : styles.break
      }`}
    >
      {status === SessionType.session ? <h2>Session</h2> : <h2>Break</h2>}

      <p className={styles.time}>{timeFormat.formatSeconds(time)}</p>
    </div>
  );
};

export default Clock;
