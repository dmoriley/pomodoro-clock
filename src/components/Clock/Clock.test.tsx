import React from 'react';
import Clock from './Clock';
import { render, screen } from '@testing-library/react';
import { SessionType } from '../../services/pomodoro';

describe('Clock', () => {
  it('renders without error', () => {
    render(<Clock time={120} status={SessionType.session} power={false} />);
  });

  it('should match snapshot when off', () => {
    const { container } = render(
      <Clock time={120} status={SessionType.session} power={false} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when on', () => {
    const { container } = render(
      <Clock time={120} status={SessionType.session} power={true} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when status is break', () => {
    const { container } = render(
      <Clock time={120} status={SessionType.break} power={true} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
