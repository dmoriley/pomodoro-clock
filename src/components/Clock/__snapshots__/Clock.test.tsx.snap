// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock should match snapshot when off 1`] = `
<div
  class="root off"
>
  <h2>
    Session
  </h2>
  <p
    class="time"
  >
    02:00
  </p>
</div>
`;

exports[`Clock should match snapshot when on 1`] = `
<div
  class="root session"
>
  <h2>
    Session
  </h2>
  <p
    class="time"
  >
    02:00
  </p>
</div>
`;

exports[`Clock should match snapshot when status is break 1`] = `
<div
  class="root break"
>
  <h2>
    Break
  </h2>
  <p
    class="time"
  >
    02:00
  </p>
</div>
`;
